src/Connection.php

```diff
@@ -122,6 +123,25 @@ protected function upsertMultiple($tableExpression, array $data, array $types =
         return $this->executeUpdate($sql, $params, $finalTypes);
     }
 
+    /**
+     * Extract ordered type list from an ordered column list and type map.
+     *
+     * @param array<int, string>     $columnList
+     * @param array<int, int|string> $types      The query parameter types.
+     *
+     * @return array<int, int>|array<int, string>
+     */
+    private function extractTypeValues(array $columnList, array $types)
+    {
+        $typeValues = [];
+
+        foreach ($columnList as $columnName) {
+            $typeValues[] = $types[$columnName] ?? ParameterType::STRING;
```

Comment:
It won't work, it should use `$types[$columnIndex]`.

Reply:
Tests are passing. Check test cases [here](https://github.com/pharako/mysql-dbal/commit/51bd4e0e7c3e8bd7f6e260592957ba9ce07b491c).